Make into a package

jeanconn opened the pull request:
Make into a package

Not sure what testing would be appropriate, but there's a change in here so it at least passes all of its own tests.

taldcroft commented on setup.py:
```diff
@@ -0,0 +1,38 @@
+
+if "--user" not in sys.argv:
+    share_path = os.path.join(sys.prefix, "share", "jobwatch")
+    web_path = os.path.join(sys.prefix, "www", "jobwatch")
```
I think there is a problem here because in a clean install of the environment this will create the `www` directory and put files there, but on HEAD the `www` directory needs to be a link to `/proj/sot/ska/www` (which is really `/data/mta4` at this time).

I'd suggest making `overlib.js` be regular package data and then copying to the destination `rootdir` if needed at runtime.

jeanconn replied:
I was thinking in a clean install intended to be a production copy, that one makes the `www` directory to start (like I did for the new HEAD ska3/flight) and this works.  In a clean install without `www` this makes it and then doesn't work, but I don't think we care for non-production environments.  In the already-set-up production environment, this should put the file we want in the right place. I'm not seeing the problem.

Could also use a post-install hook for conda, or the runtime copy you suggestion, but didn't seem necessary.

taldcroft replied:
Last time I checked, conda won't do a clean install into a directory that exists and has stuff in it. For the shiny production install you did the link *after* the install. Is that different now? 

In any case, we should minimize the external data `data_files` to the extent possible. Basically `overlib.js` is just another output file analogous to, and required by, the generated content. It's worth noting that the template file hardcodes the location of overlib.js at `../overlib.js`, so this will be broken unless the output `rootdir` happens to be a subdirectory of `web_path`. Instead let's do this right and have `overlib.js` live in `rootdir`.

jeanconn replied:
I don't know about "clean install into a directory that exists and has stuff in it" in this context.  I had done a test install into a bare ska, and it made the 'www' directory and stuck the overlib.js in it.  Then I removed the package, made the 'www' a link to a temp directory and repeated that test without issue.  I don't think we'd care about deleting a previous www/jobwatch directory as long as we didn't have to mess with `www` itself?

I'm not arguing if you want to minimize data_files. But along those lines... are you actually looking at the overlib.js content?  I confess I've never been to the skawatch index page where that works.

taldcroft replied:
The real goal is to make the fresh install process as simple and clean as possible, and have zero post-install steps for particular packages. The only thing we absolutely need now is the `data` link. 

What you seem to be proposing is a post-install step for this package to remove the package-created `www` and replace it with the link, and then do some by-hand copying of `overlib.js` to put it back into the right place. Then you end up having part of the configured Ska3 flight distribution living on /data/mta4. This is really not ideal.

So please just do what I suggested.

jeanconn replied:
Sure.  

PS I was not proposing that.  I was mentioning a) it seemed to work without post-install steps (though yes, in a clean install you'd now make 'www' link before installing the ska conda packages instead of after). b) if it needed post-install steps I'd do them as conda post-install.sh c) I still don't think anybody uses the overlib data.

taldcroft commented on jobwatch/hourly_watch.py:
```diff
@@ -21,6 +21,23 @@
+                        help='Send email report')
+    parser.add_argument('--loud',
+                        action='store_true',
+                        help='Send email report')
```
Copy/paste (probably pre-existing?) error.

taldcroft commented on jobwatch/hourly_watch.py:
```diff
@@ -21,6 +21,23 @@
 
 
+def get_options():
+    parser = argparse.ArgumentParser(description='Replan Central monitor')
```
The description is a leftover, apparently.

jeanconn replied:
Well, it is still *basically* a replan central monitor, but we've since added an "is kadi up" check and called it the hourly watch so...

taldcroft reviewed (changes requested):
Looks generally OK apart from the one comment about overlib.

About testing, can you set this up in a dedicated testing Ska3 and run it in parallel for a week or two (maybe without the --email flag)? I'm not sure how hard this would be out of the box, but we do need to be very confident that it will work the first time when we promote.

jeanconn commented:
With regard to testing, I don't quite see the need for a soak period?  This doesn't really have different behaviors over time.

jeanconn commented on jobwatch/jobwatch.py:
```diff
@@ -35,7 +35,7 @@ def __init__(self, task, filename,
         self.filetime = None
         self.filedate = None
-
+        self.type = 'Job'
```
This hack was to make a test pass, but I think I should probably just update the code in reporting that looks for a  "jw.type" and skip if type is missing.

taldcroft replied:
Does it work if you assign `None`. This would be a bit more agnostic.

jeanconn replied:
No, None didn't work.  And setting it to 'Job' seemed to just override the rest of the types.  I didn't have the brainpower just now to figure out the inheritance and just updated the logic in the html report method for now.

taldcroft commented:
The soak period will naturally happen between when you start the job going and we are ready with a release. I'd just like to see an end-to-end test for a few days because it is so easy to make a little mistake in packaging (I've done this many times).

jeanconn commented:
@taldcroft if you were interested in installing this in the next ska3 update but also running on the side, I think we would have a little bit of a collision issue for data/skawatch3 and www/skawatch3 .  Did you want to, say, reclaim `skawatch` instead of `skawatch3`?

jeanconn commented:
Actually, since we win nothing (no updated checks or fixes in here) by putting this PR in skare3 2021.1, I think we should just hold this one.

jeanconn commented on jobwatch/jobwatch.py:
```diff
@@ -174,16 +173,19 @@ def set_report_attrs(jobwatches):
-        if i_jw == 0 or jw.type != jobwatches[i_jw - 1].type:
-            jw.span_cols_text = jw.type
+
+        this_type = getattr(jw, 'type', 'Job')
+        last_type = getattr(jobwatches[i_jw -1], 'type', 'Job')
+        if i_jw == 0 or this_type != last_type:
+            jw.span_cols_text = this_type
```
I'm not sure if jw type should be a property, but this more explicit handling of the attr at least seemed not to break much (related to comment at https://github.com/sot/jobwatch/pull/41#discussion_r551679164) .

jeanconn commented:
OK.  I think I've addressed the comments and we probably don't care about fixing jobwatch type inheritance or whatever, so merging despite the "changes requested" bit.
